src/dns/server.rs

```diff
@@ -333,34 +334,43 @@ impl Store {
                 let search_name_str = search_name.to_string().into();
                 search_name.set_fqdn(true);
 
-                // First, lookup the host as a service.
-                if let Some(services) = state.services.get_by_host(&search_name_str) {
-                    // We found a match. We always return `Some` result, even if there
-                    // are zero records returned.
-
+                let service = state
+                    .services
+                    .get_by_host(&search_name_str)
+                    .iter()
+                    .flatten()
+                    // Remove things without a VIP, unless they are Kubernetes headless services.
+                    // This will trigger us to forward upstream.
+                    // TODO: we should have a reliable way to distinguish these. In sidecars, we use
+                    // `svc.Attributes.ServiceRegistry`, but we don't pass anything similar over WDS.
+                    // For now, checking the domain is good enough.
+                    // This does mean a `.svc.cluster.local` ServiceEntry will use these semantics, but
+                    // its better than *ALL* ServiceEntry doing this
+                    .filter(|service| {
+                        // Domain will be like `.svc.cluster.local.` (trailing .), so ignore the last character.
+                        let domain = self.svc_domain.to_utf8();
+                        !service.vips.is_empty()
+                            || service.hostname.ends_with(&domain[..domain.len() - 1])
```

Comment:
Should we use strip_suffix(...).expect("the svc domain must have a trailing '.'") which is a firmer assertion of the form of the svc_domain since malformed will cause a panic but won't quietly behave badly if our assumptions were false.